### PR TITLE
Factors outbound http

### PR DESCRIPTION
**Pull request by lann**
Pending #2703, and probably some tests (?)

### CODE REVIEW

File: crates/factor-outbound-http/src/lib.rs
Comment (lann) [inline crates/factor-outbound-http/src/lib.rs:140]: I thought I was going to need more fields on `Intercepted`. Could probably flatten them into `intercept` args.

File: crates/factor-outbound-http/src/wasi.rs
Comment (lann) [inline crates/factor-outbound-http/src/wasi.rs:115]: This skips a lot of logic from the current impl but afaict `Authority` enforces everything we care about.

File: crates/factor-outbound-http/src/wasi.rs
Comment (lann) [inline crates/factor-outbound-http/src/wasi.rs:139]: This is significantly reorganized but hopefully easier to follow.

File: crates/trigger-http2/src/lib.rs
Comment (lann) [inline crates/trigger-http2/src/lib.rs:299]: TODID

File: crates/trigger-http2/src/outbound_http.rs
Comment (lann) [inline crates/trigger-http2/src/outbound_http.rs:50]: I don't really understand whats going on here (copied from spin-trigger-http) but it compiles!

File: crates/trigger-http2/src/outbound_http.rs
Comment (lann) [inline crates/trigger-http2/src/outbound_http.rs:34]: @itowlson Porting service chaining to factors

File: crates/trigger-http2/src/server.rs
Comment (lann) [inline crates/trigger-http2/src/server.rs:146]: This part of the diff is mostly indentation.

File: crates/trigger-http2/src/server.rs
Comment (lann) [inline crates/trigger-http2/src/server.rs:130]: This diff chunk is mostly whitespace (dedent).

File: crates/factor-outbound-http/src/lib.rs
Comment (rylev) [inline crates/factor-outbound-http/src/lib.rs:80]: Are we sure we don't want to just silently overwrite the interceptor?

File: crates/factor-outbound-http/src/lib.rs
Comment (rylev) [inline crates/factor-outbound-http/src/lib.rs:125]: Nit: some docs on the trait itself (if for nothing else than to help with IDE hover overlays).

File: crates/factor-outbound-http/src/wasi.rs
Comment (rylev) [inline crates/factor-outbound-http/src/wasi.rs:86]: Are we asking for issues if `host` is `""`? I guess this _should_ return the default tls config, and everything will just work, but it does seem a bit round about, and I worry things might get confusing with future code changes.

File: crates/factor-outbound-http/src/wasi.rs
Comment (rylev) [inline crates/factor-outbound-http/src/wasi.rs:152]: I'd like to figure this out sooner rather than later since we're requiring a fairly Spin CLI specific logic (e.g., the use of `terminal::warn`) on every user of this factor. It would be nice to ensure we can handle this outside of the factor.

File: crates/factor-outbound-http/src/lib.rs
Comment (rylev) [inline crates/factor-outbound-http/src/lib.rs:133]: Instead of this returning an `Option` and sneaking the origin through the request extensions, I wonder if instead we could make `OutboundHttpInterceptor` generic and have:

```rust
pub trait OutboundHttpInterceptor<T>: Send + Sync {
    fn intercept(&self, intercepted: Intercepted) -> Intercepted<T>;
}

enum InterceptionResult<T> {
    Intercepted(HttpResult<HostFutureIncomingResponse>),
    NotIntercepted(T)
}
```

Then in `impl` the origin is stored in `InterceptionResult::NotIntercepted` which eliminates the implicit storing of the information in the request.

There's no need to block on this idea, we can also try it out after we merge this.

File: crates/factor-outbound-http/tests/factor_test.rs
Comment (rylev) [inline crates/factor-outbound-http/tests/factor_test.rs:42]: Nit:
```
future_resp.unwrap_ready().unwrap_err()
```

File: crates/factor-outbound-networking/src/lib.rs
Comment (rylev) [inline crates/factor-outbound-networking/src/lib.rs:168]: I know this was just moved, but why doesn't `?` work?

File: crates/trigger-http2/src/outbound_http.rs
Comment (rylev) [inline crates/trigger-http2/src/outbound_http.rs:26]: I'm surprised to see this remain in the trigger. I had been thinking that factors would allow us to move this logic (which doesn't seem to be trigger specific) closer to the outbound-http handler code.

**Review by rylev — CHANGES_REQUESTED** [review]
Mostly looks pretty good. The only thing I'd like to figure out before we merge is the question of where the request interception impl should live. I was surprised to see it in the trigger crate when the logic appears to not be trigger dependent.

File: crates/trigger-http2/src/outbound_http.rs
Comment (rylev) [reply crates/trigger-http2/src/outbound_http.rs:26]: Looking more deeply into this, I think I understand why at least *some* of this has to be in the trigger. The interceptor is meant as a way for the trigger to inject custom logic on each request. In this case, essentially the HTTP trigger's interceptor is responsible for the following things:
1. Establishing the `SelfRequestOrigin` so that self requests can be routed properly.
  1a. The factor has an implicit assumption that somewhere someone is adding a `SelfRequestOrigin` to the request extensions. If no one does, requests to `self` will fail. 
2. Figuring out whether the request is a a service chained request. If it is a service chaining request, then routing that back through the HTTP trigger. 

Both seem to be very heavily tied to the specific HTTP trigger though I'm realizing now, I'm not sure we ever really rationalized these features outside of the HTTP trigger. What does a self request or a service chained request mean when the app is invoked outside of the HTTP trigger? 

As the code is currently written, self requests seem *indirectly* to not be a feature outside of the HTTP trigger. If the factor is invoked outside of an HTTP trigger and no origin is set on the request extensions, then the outgoing request will fail. If a trigger wants to opt into supporting self requests, they have to do so in a very indirect way (by setting the `SelfRequestOrigin` on the request extensions) - this is certainly less than ideal... The outbound-http factor code is very much aware of this feature, but exposes it to embedders in a very indirect way. 

Service chaining, on the other hand, seems to be explicitly an HTTP trigger only feature. If "spin.internal" is used in an outbound request in any other trigger, that request will be routed to "spin.internal" as if it were any other DNS based authority (I guess let's hope that the user is running Spin on a network where `spin.internal` doesn't route to anything malicious...). The outbound-http factor knows nothing about this feature, and it is implemented purely within the HTTP trigger.

Comment (lann) [reply crates/trigger-http2/src/outbound_http.rs:26]: I explored a couple of different approaches here but decided to follow the existing semantics for pretty much the reasons you mention:

- "origin" is inherently somewhat trigger-specific; we _could_ have a universal e.g. runtime config option to set/override it for any trigger type, but that seemed out of scope for a refactor

- service chaining could be implemented for other trigger types but would require a large chunk of the http trigger be included in the outbound http factor which again seems out of scope

File: crates/factor-outbound-http/src/lib.rs
Comment (lann) [reply crates/factor-outbound-http/src/lib.rs:80]: Pretty sure. Since this is used with dependency inversion the end developer could be surprised when composing two factors that call this if one of them silently loses its interceptor.

The real solution would be to allow multiple interceptors but that is a little complicated in rust so I punted.

File: crates/factor-outbound-http/src/lib.rs
Comment (lann) [reply crates/factor-outbound-http/src/lib.rs:133]: The enum result type is probably a good idea regardless, but I personally think the extension is fine here 🤷

File: crates/factor-outbound-http/src/wasi.rs
Comment (lann) [reply crates/factor-outbound-http/src/wasi.rs:86]: Well, `host` could also be present-but-empty I think? Or do your recent changes make that impossible?

File: crates/trigger-http2/src/outbound_http.rs
Comment (rylev) [reply crates/trigger-http2/src/outbound_http.rs:26]: All of that sounds good to me - I have another comment about how I don't love how we're feeding the origin information back into the factor, but that's an implementation detail we can change after we merge this PR if we want.

File: crates/factor-outbound-http/src/wasi.rs
Comment (rylev) [reply crates/factor-outbound-http/src/wasi.rs:86]: I'm actually not really sure. I *think* from a logic point of view this is probably fine. Then again, just because something works doesn't mean it's the right implementation. I personally would likely feel better with a more verbose and explicit implementation, but I won't die on this hill 😄.

File: crates/factor-outbound-http/src/lib.rs
Comment (rylev) [reply crates/factor-outbound-http/src/lib.rs:133]: It's a balance - the reason I'm not a fan of extension approach is that it's a very roundabout public API. In order to unlock support for self, users of this factor need to know that they have to somehow add a `SelfRequestOrigin` to the request extensions. It works, and we certainly can make sure this documented somewhere, but it's not a particularly ergonomic or intuitive API. 

I get that this library is unlikely to be used by a large number of people, so if we can't be super ergonomic/easy to understand, that's not the end of the world, but for my future self's sake, I'd love for us to try to make this more straight forward.

Comment (lann) [reply crates/factor-outbound-http/src/lib.rs:133]: Added return enum `InterceptOutcome`

File: crates/factor-outbound-http/tests/factor_test.rs
Comment (lann) [reply crates/factor-outbound-http/tests/factor_test.rs:42]: The `Ok` side doesn't impl `Debug`.

File: crates/factor-outbound-networking/src/lib.rs
Comment (lann) [reply crates/factor-outbound-networking/src/lib.rs:168]: `err` is an `Arc<Error>` here

**Comment by lann** [comment]
Addressed some of the feedback and added a couple of basic smoke tests.

**Review by rylev — APPROVED** [review]

File: crates/runtime-config/src/lib.rs
Comment (lann) [inline crates/runtime-config/src/lib.rs:1]: @rylev added this

**Pull request merged**
